pkg/model/azuremodel/api_loadbalancer.go

```diff
@@ -38,5 +45,122 @@ func (b *APILoadBalancerModelBuilder) Build(c *fi.ModelBuilderContext) error {
 		return nil
 	}
 
-	return errors.New("using loadbalancer for API server is not yet implemented in Azure")
+	lbSpec := b.Cluster.Spec.API.LoadBalancer
+	if lbSpec == nil {
+		// Skipping API LB creation; not requested in Spec
+		return nil
+	}
+
+	switch lbSpec.Type {
+	case kops.LoadBalancerTypeInternal:
+		// OK
+	case kops.LoadBalancerTypePublic:
+		// TODO: Implement creating public ip and attach to public loadbalancer
+		return fmt.Errorf("only internal loadbalancer for API server is implemented in Azure")
+	default:
+		return fmt.Errorf("unhandled LoadBalancer type %q", lbSpec.Type)
+	}
+
+	// Create LoadBalancer for API ELB
+	var lb *azuretasks.LoadBalancer
+	{
+		lb = &azuretasks.LoadBalancer{
+			Name:          fi.String(b.NameForLoadBalancer()),
+			Lifecycle:     b.Lifecycle,
+			ResourceGroup: b.LinkToResourceGroup(),
+			Tags:          map[string]*string{},
+		}
+
+		switch lbSpec.Type {
+		case kops.LoadBalancerTypeInternal:
+			lb.RequirePublicIP = to.BoolPtr(false)
+			subnet, err := b.subnetForLoadBalancer()
+			if err != nil {
+				return err
+			}
+			lb.Subnet = b.LinkToAzureSubnet(subnet)
+		case kops.LoadBalancerTypePublic:
+			lb.RequirePublicIP = to.BoolPtr(true)
+		default:
+			return fmt.Errorf("unknown load balancer Type: %q", lbSpec.Type)
+		}
+
+		c.AddTask(lb)
+	}
+
+	if dns.IsGossipHostname(b.Cluster.Name) || b.UsePrivateDNS() {
+		lb.ForAPIServer = true
+	}
+
+	return nil
+}
+
+type scoredSubnet struct {
+	score  int
+	subnet *kops.ClusterSubnetSpec
+}
+
+type ByScoreDescending []*scoredSubnet
+
+func (a ByScoreDescending) Len() int      { return len(a) }
+func (a ByScoreDescending) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a ByScoreDescending) Less(i, j int) bool {
+	if a[i].score != a[j].score {
+		// ! to sort highest score first
+		return !(a[i].score < a[j].score)
+	}
+	// Use name to break ties consistently
+	return a[i].subnet.Name < a[j].subnet.Name
+}
+
+// subnetForLoadBalancer returns the subnet the loadbalancer will use.
+func (c *AzureModelContext) subnetForLoadBalancer() (*kops.ClusterSubnetSpec, error) {
+	var subnets []*kops.ClusterSubnetSpec
+	for i := range c.Cluster.Spec.Subnets {
+		subnet := &c.Cluster.Spec.Subnets[i]
+
+		if c.Cluster.Spec.API.LoadBalancer.Type == kops.LoadBalancerTypeInternal {
+			if subnet.Type != kops.SubnetTypePrivate {
+				continue
+			}
+		}
+
+		subnets = append(subnets, subnet)
+	}
+
+	if len(subnets) == 0 {
+		return nil, nil
+	}
+	if len(subnets) == 1 {
+		return subnets[0], nil
+	}
+
+	migSubnets := sets.NewString()
```

Comment:
Could you please add some comment and explain the logic? Couldn't follow..

Reply:
i got this function from the aws load balancer file. but it basically searches through all the subnets for the master instance group subnets and if found, scored with a value of 1. then at the end it sorts the subnets from greatest to least and the first element in the array is returned. this way it should be in a master subnet. 

i'm fine with if you'd rather just have the first master subnet found be returned and the internal loadbalancer be assigned to that subnet. if you think thats a better way to do it?

Reply:
Ah, I see. Thanks for the explanation! I'll defer to other reviewers who know the code better.

Reply:
I think the reason this complexity exists in AWS is because kops needs to choose a subnet per availability zone to use in its load balancer.

Do Azure load balancers support multiple subnets? if not then we could reduce this down to just returning the first master subnet as you mentioned.